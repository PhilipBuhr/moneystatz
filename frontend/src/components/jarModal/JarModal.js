import React from "react";
import { connect } from "react-redux";
import './JarModal.css';
import Modal from "../commons/Modal";
import { closeAddJar, deleteJar, selectJar, submitJar } from "../../actions/transactionActions";

export class JarModal extends React.Component {

    render() {
        if (!this.props.active) {
            return null;
        }
        return (
            <Modal onClose={this.onClose} onSubmit={this.onSubmit} onDelete={this.onDelete}>
                <div className="AddJarModal-body">
                    <div>Jar</div><input value={this.props.jar.name} onChange={event => this.onChange(event, 'name')}/>
                    <div>Type</div>
                    <select value={this.props.jar.type} onChange={event => this.onChange(event, 'type')}>
                        <option value="income">Income</option>
                        <option value="expense">Expense</option>
                    </select>
                    <div>Order</div><input value={this.props.jar.order} onChange={event => this.onChange(event, 'order')}/>
                </div>
            </Modal>
        );
    }

    onChange = (event, property) => {
        const newJar = { ...this.props.jar };
        newJar[property] = event.target.value;
        this.props.change(newJar);
    };

    onSubmit = () => {
        this.props.submit(this.props.jar);
    };

    onDelete = () => {
        this.props.delete(this.props.jar)
    };

    onClose = () => {
        this.props.close();
    };
}

const mapStateToProps = state => ({
    active: !!state.dateState.selectedJar,
    jar: state.dateState.selectedJar
});

const mapDispatchToProps = dispatch => ({
    close: () => dispatch(closeAddJar()),
    change: jar => dispatch(selectJar(jar)),
    submit: jar => dispatch(submitJar(jar)),
    delete: jar => dispatch(deleteJar(jar))
});

export default connect(mapStateToProps, mapDispatchToProps)(JarModal);
