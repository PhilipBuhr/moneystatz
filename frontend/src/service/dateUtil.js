export const format = date => date.toISOString().substring(0, 10);

export const parse = dateString => {
    const parts = dateString.split('-');
    const year = parseInt(parts[0]);
    const month = parseInt(parts[1]);
    const day = parseInt(parts[2]);
    return new Date(Date.UTC(year, month - 1, day));
};
