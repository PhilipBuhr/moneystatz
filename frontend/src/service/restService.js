import axios from 'axios';

export class RestService {
    constructor(host='localhost', port = 8000) {
        this.url = `http://${host}:${port}`
    }

    get(path, params) {
        let param_string = this.buildUrlParams(params);
        return axios.get(`${this.url}/${path}${param_string}`);
    }

    post(path, data) {
        const options = {
            headers: {'Content-Type': 'application/json'}
        };
        return axios.post(`${this.url}/${path}`, data, options);
    }

    delete(path, id) {
        return axios.delete(`${this.url}/${path}/${encodeURIComponent(id)}`);
    }

    buildUrlParams(params) {
        const url_params = [];
        for (let key in params) {
            if (params.hasOwnProperty(key)) {
                key = encodeURIComponent(key);
                let value = encodeURIComponent(params[key]);
                url_params.push(`${key}=${value}`);
            }
        }
        return url_params ? `?${url_params.join('&')}` : '';
    }
}

export default RestService;
