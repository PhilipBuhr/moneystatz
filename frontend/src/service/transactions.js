export class Transactions {
    constructor(jars) {
        this.jars = jars;
        this.groupedTransactions = {};
        this.jars.forEach(jar => this.groupedTransactions[jar] = [])
    }

    static parse(json) {
        let transactions = new Transactions(json['jars']);
        json['transactions'].forEach(transaction => transactions.push(transaction));
        return transactions;
    }

    push(transaction) {
        let jar = transaction['jar'];
        this.groupedTransactions[jar].push(transaction);
    }

    get(jar) {
        let transactions = this.groupedTransactions[jar];
        if (transactions) {
            return transactions;
        }
        return [];
    }

    maxJarSize() {
        return this.jars
            .map(jar => this.get(jar))
            .map(transactions => transactions.length)
            .reduce((max, size) => Math.max(max, size), 0);
    }

    toTotals() {
        return this.jars.map(jar => {
            const total = this.get(jar)
                .map(transaction => transaction.amount)
                .reduce((sum, amount) => sum + amount, 0);
            return {jar: jar, total: total}
        })
    }
}
