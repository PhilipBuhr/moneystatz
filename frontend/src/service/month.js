import {format} from "./dateUtil";

export class Month {
    constructor(month, year) {
        while (month > 12) {
            month -= 12;
            year += 1;
        }
        while (month < 1) {
            month += 12;
            year -= 1;
        }
        this.month = month;
        this.year = year;
    }

    static from(date) {
        return new Month(date.getMonth() + 1, date.getFullYear());
    }

    getPrevious() {
        return new Month(this.month - 1, this.year);
    }

    getNext() {
        return new Month(this.month + 1, this.year)
    }

    getFirstAsString() {
        let first = new Date(Date.UTC(this.year, this.month -1 , 1));
        return format(first);
    }

    getLastAsString() {
        const last = new Date(Date.UTC(this.year, this.month, 0));
        return format(last);
    }

    getMonthName() {
        const date = new Date(Date.UTC(this.year, this.month - 1, 1));
        return date.toLocaleString('default', { month: 'long' });
    }
}
