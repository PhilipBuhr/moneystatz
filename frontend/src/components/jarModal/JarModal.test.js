import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect'
import { JarModal } from "./JarModal";
import { Month } from '../../service/month';


describe('JarModal', () => {
    let onClose;
    let onSubmit;

    beforeEach(() => {
        onClose = jest.fn();
        onSubmit = jest.fn();
    });

    it('should render', () => {
        const { container } = render(
            <JarModal
                active={true}
                month={new Month(3, 2020)}
                close={onClose}
                submit={onSubmit}
            />);
        expect(container).toMatchSnapshot('JarModalSnapshot');
    });

    it('should close only on background click', () => {
        const { container } = render(
            <JarModal
                active={true}
                month={new Month(3, 2020)}
                close={onClose}
                submit={onSubmit}
            />);

        fireEvent.click(container.querySelector('.Modal-modal'));

        expect(onClose).not.toHaveBeenCalled();

        fireEvent.click(container.querySelector('.Modal-container'));

        expect(onClose).toHaveBeenCalled();
    });

    it('should close on close button click', () => {
        render(
            <JarModal
                active={true}
                month={new Month(3, 2020)}
                close={onClose}
                submit={onSubmit}
            />);

        fireEvent.click(screen.getByAltText('Cancel'));

        expect(onClose).toHaveBeenCalled();
    });

    it('should submit new Jar', () => {
        const { container } = render(
            <JarModal
                active={true}
                month={new Month(3, 2020)}
                close={onClose}
                submit={onSubmit}
            />);

        fireEvent.change(container.querySelector('input'), { target: { value: 'jarName' } });
        fireEvent.click(screen.getByAltText('Submit'));

        expect(onSubmit).toHaveBeenCalledWith('jarName', {month: 3, year: 2020});
    });
});
