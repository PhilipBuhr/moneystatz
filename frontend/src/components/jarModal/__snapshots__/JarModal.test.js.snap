// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JarModal should render: JarModalSnapshot 1`] = `
<div>
  <div
    class="Modal-container"
  >
    <div
      class="Modal-modal"
    >
      <div
        class="Modal-body"
      >
        <div
          class="AddJarModal-body"
        >
          <div>
            Jar
          </div>
          <input
            value=""
          />
        </div>
      </div>
      <div
        class="Modal-button-box"
      >
        <button
          class="Modal-button submit"
        >
          <img
            alt="Submit"
            class="Modal-icon-button"
            src="check-24px.svg"
          />
        </button>
        <button
          class="Modal-button close"
        >
          <img
            alt="Cancel"
            class="Modal-icon-button"
            src="close-24px.svg"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
